### PR TITLE
TechDocs: Support AWS S3 as TechDocs external storage

### LINKED ISSUE #3714: TechDocs: Support AWS S3 as TechDocs external storage (to store and read generated documentation sites)
## Feature Suggestion

In the [TechDocs Architecture](https://backstage.io/docs/features/techdocs/architecture), users should be able to use any kind of storage system to store their generated documentation site.

Currently, Google Cloud Storage and Local filesystem are supported. Add support for Amazon S3 buckets as well.

## Possible Implementation

TechDocs has a main [Publisher](https://github.com/backstage/backstage/blob/master/packages/techdocs-common/src/stages/publish/publish.ts) factory class which initialises a publisher based on app's config `techdocs.publisher.type` [in the backend](https://github.com/backstage/backstage/blob/767b679f4388b9296d2f18af12f4f7ccc994b38b/packages/backend/src/plugins/techdocs.ts#L52).

A TechDocs publisher is responsible for two things (two-way communication between `techdocs-backend` and the storage)
1. Publish generated static files to a storage. (Only if the `techdocs.builder` is set to `local` which means Backstage is supposed to locally generate sites from source markdown files and publish to storage)
2. Read files from storage when users visit a TechDocs page (TechDocs requests `techdocs-backend` which uses the Publisher's APIs).

Add a new `amazonS3.ts` publisher [here](https://github.com/backstage/backstage/tree/master/packages/techdocs-common/src/stages/publish) similar to [`googleStorage.ts` publisher](https://github.com/backstage/backstage/blob/master/packages/techdocs-common/src/stages/publish/googleStorage.ts). The S3 publisher needs to implement [`PublisherBase`](https://github.com/backstage/backstage/blob/767b679f4388b9296d2f18af12f4f7ccc994b38b/packages/techdocs-common/src/stages/publish/types.ts#L40-L64) abstract class and its four methods. Use Google Storage as an example since S3 is going to be very similar.

Some additional checklist
- Add new configs under `techdocs.publisher`. Update config reference documentation. https://backstage.io/docs/features/techdocs/configuration
- Add steps about how to use Amazon S3, similar to GCS here https://backstage.io/docs/features/techdocs/using-cloud-storage
- Add changesets for the plugin and packages (packages/techdocs-common and plugins/techdocs-backend)
- Make sure to add tests and mocks. :) (Use the PR below for example on GCS.)

## Context

PR which added Google Cloud Storage support - https://github.com/backstage/backstage/pull/3432

Feel free to ask questions/clarifications here. You can also use `#docs-like-code` channel on Discord for any help (we can pair program together if that helps).

### PR DESCRIPTION
**Pull request by ayshiff**
## Hey, I just made a Pull Request!

Closes #3714

This PR adds the feature to use AWS S3 as TechDocs external storage.

- Added a new `awsS3.ts` publisher [Link](https://github.com/backstage/backstage/blob/174fede6f08fc4234dae748116cf6b388b03acbc/packages/techdocs-common/src/stages/publish/awsS3.ts)
- Added tests and mocks [Link](https://github.com/backstage/backstage/blob/174fede6f08fc4234dae748116cf6b388b03acbc/packages/techdocs-common/src/stages/publish/awsS3.test.ts)
- Added steps about how to use AWS S3 [Link](https://github.com/backstage/backstage/blob/174fede6f08fc4234dae748116cf6b388b03acbc/docs/features/techdocs/using-cloud-storage.md)
- Updated config reference documentation [Link](https://github.com/backstage/backstage/blob/174fede6f08fc4234dae748116cf6b388b03acbc/docs/features/techdocs/configuration.md) 

#### :heavy_check_mark: Checklist

- [X] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [X] Added or updated documentation
- [X] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)

### CODE REVIEW

File: .changeset/fast-colts-cross.md
Comment (OrkoHunter) [inline .changeset/fast-colts-cross.md:7]: There seems to be two exactly same changesets.

File: app-config.yaml
Comment (OrkoHunter) [inline app-config.yaml:79]: 'Or' might be better?

```suggestion
    type: 'local' # Alternatives - 'googleGcs' or 'awsS3'. Read documentation for using alternatives.
```

File: docs/features/techdocs/configuration.md
Comment (OrkoHunter) [inline docs/features/techdocs/configuration.md:55]: I figure we could split this into two lines.
```suggestion
    # When set to 'googleGcs', techdocs-backend will use a Google Cloud Storage Bucket to store generated documentation files.
    # When set to 'awsS3', techdocs-backend will use an Amazon Web Service (AWS) S3 bucket to store generated documentation files.

```

File: docs/features/techdocs/using-cloud-storage.md
Comment (OrkoHunter) [inline docs/features/techdocs/using-cloud-storage.md:140]: Some language suggestions. :)
```suggestion
Create an **administrator user** account `ADMIN_USER` and grant it administrator
privileges by attaching a user policy giving the account **full access**.  
Note down the Admin User credentials and IAM User Sign-In URL as you will need
to use this information in the next step.
```

File: packages/techdocs-common/src/stages/publish/awsS3.ts
Comment (OrkoHunter) [inline packages/techdocs-common/src/stages/publish/awsS3.ts:98]:
```suggestion
      // Note: S3 manages creation of parent directories if they do not exist.
```

File: packages/techdocs-common/src/stages/publish/awsS3.ts
Comment (OrkoHunter) [inline packages/techdocs-common/src/stages/publish/awsS3.ts:215]: Since we only care if the file exists, instead of getting the object, we can do [`.headObject`](https://docs.aws.amazon.com/AmazonS3/latest/API/API_HeadObject.html) or just `check if object exists` if possible.

File: plugins/techdocs-backend/src/service/router.ts
Comment (OrkoHunter) [inline plugins/techdocs-backend/src/service/router.ts:152]:
```suggestion
            // The delay here is to make sure GCS/AWS/etc. registers newly uploaded files which is usually <1 second
```

**Review by OrkoHunter — COMMENTED** [review]
Looking great so far @ayshiff ! Kudos for your work on this and supporting documentation as well.

Leaving some comments. Will take another look very soon. 🎉

File: .changeset/fast-colts-cross.md
Comment (OrkoHunter) [inline .changeset/fast-colts-cross.md:4]: 1. These should be patch changes. Since we are still v0.x.y, minor = breaking changes, patch = new features and bug fixes.
2. `@backstage/plugin-techdocs` can be omitted since we are not updated that package.

**Review by andrewthauer — COMMENTED** [review]
Excited this is being added.

File: docs/features/techdocs/configuration.md
Comment (ayshiff) [reply docs/features/techdocs/configuration.md:79]: I agree with you. 
At first I wanted to put the credentials as environment variables:

```yml
credentials:
    aws_access_key_id: ACCESS_KEY_ID
    aws_secret_access_key: SECRET_ACCESS_KEY
```

But to keep some consistency with the configuration for GCS I thought it might be interesting to keep the same credentials file logic. 

But since it is not very relevant for aws we can perhaps replace file credentials by environment variables 👍

@OrkoHunter I would like to know your opinion on that.

Comment (OrkoHunter) [reply docs/features/techdocs/configuration.md:79]: > Curious, can the credentials be set with just environment variables?

Yes! We can use the contents of the "**credentials.json" as the value for the environment variable (e.g. `TECHDOCS_AWSS3_CREDENTIALS`). And set
```
awsS3:
  credentials:
    $env: TECHDOCS_AWSS3_CREDENTIALS
```

I have used this with GCS. But I see that it will be nice to write this ^ in the `app-config.yaml` comments and the Config reference docs.

Comment (andrewthauer) [reply docs/features/techdocs/configuration.md:79]: Wouldn't the environment variable need to be a JSON string in this case? I'm personally not a big fan of managing them that way.

Comment (OrkoHunter) [reply docs/features/techdocs/configuration.md:79]: @andrewthauer I agree with you that JSON strings are not pretty in environment variables.

Here is my concern. With GCS, their credentials JSON looks like this
```json
{
  "type": "",
  "project_id": "",
  "private_key_id": "",
  "private_key": "-----BEGIN PRIVATE KEY-----<Insert 1600 characters>-----END PRIVATE KEY-----\n",
  "client_email": "",
  "client_id": "",
  "auth_uri": "",
  "token_uri": "",
  "auth_provider_x509_cert_url": "",
  "client_x509_cert_url": ""
}
```

Even if we split these out into individual fields, the `private_key` will still be a huge JSON string. So, I do not see much of an improvement here over the existing ways i.e.
1. Copy over the credentials file as a huge `$env` variable.
2. (Kubernetes) Inject the `$file` using Kubernetes secrets management. https://kubernetes.io/docs/concepts/configuration/secret/#overview-of-secrets
3. ...other typical ways to make a secret file available on a server.

That being said, if AWS allows smaller alphanumeric token-type secrets that can be generated and set as environment variables, I too would prefer that over the huge JSON string. It's just that, GCS is holding us back with how it authenticates service accounts.

Comment (andrewthauer) [reply docs/features/techdocs/configuration.md:79]: Ok, in that case, then yeah, I'm not sure it makes sense to mimic GCP for this then. I wonder if these could be a pass through where the child config fields are just turned into an object as is. This would potentially allow you to use both file or individual fields.

For the AWS storage, it might also be nice to support temporary security tokens with STS. This could be done in a subsequent PR, but it would change the config required for security credentials a bit.

Comment (ayshiff) [reply docs/features/techdocs/configuration.md:79]: From what I understand we would have an environment variable `TECHDOCS_AWSS3_CREDENTIALS` with the content of the `credentials.json` stringified but wouldn't it be more relevant to set the credentials individually ?



```yml
awsS3:
  credentials:
    aws_access_key_id:
        $env: TECHDOCS_AWSS3_ACCESS_KEY_ID_CREDENTIAL
    aws_secret_access_key:
        $env: TECHDOCS_AWSS3_SECRET_ACCESS_KEY_CREDENTIAL
```

Comment (OrkoHunter) [reply docs/features/techdocs/configuration.md:79]: @ayshiff Yes, please go ahead if you think it's an easier process to directly set env variables for S3. I'll be looking forward to the documentation you have written to test the process for myself. Thank you!

Comment (andrewthauer) [reply docs/features/techdocs/configuration.md:79]: Actually the keys themselves should be camelCases by convention:

I would suggest:

```yaml
awsS3:
  credentials:
    accessKeyKd:
      $env: TECHDOCS_AWSS3_ACCESS_KEY_ID_CREDENTIAL
    secretAccessKey:
      $env: TECHDOCS_AWSS3_SECRET_ACCESS_KEY_CREDENTIAL
```

To extend this for `STS` in the future it might look something like this:

```yaml
awsS3:
  credentials:
    mode: 'iam-role'
    region:
      $env: AW_S3_REGION
```

File: docs/features/techdocs/using-cloud-storage.md
Comment (OrkoHunter) [inline docs/features/techdocs/using-cloud-storage.md:157]: This needs to be either removed or updated with `accessKeyId` and `secretAccessKey` as described below.

**Review by OrkoHunter — APPROVED** [review]
:shipit: 

I have tested the setup with S3 and it works great! Thanks again @ayshiff for working on this.

**Comment by ayshiff** [comment]
I will fix the tests and fix the techdocs JSON schema by adding the `region` property 👍

**Comment by OrkoHunter** [comment]
Awesome!

**Comment by ayshiff** [comment]
I don't know if you have reviewed the example diagram for aws. [Link](https://github.com/backstage/backstage/blob/42fe286c0270df61206b71a585727a2a77b4d243/docs/assets/techdocs/aws-s3.png)

I used [excalidraw](https://excalidraw.com/) to draw it but saw that the other graphs were made with [draw.io](https://app.diagrams.net/).
Do you think it's better that I redo the graph with draw.io to keep the documentation consistent ?

<p align="center">
aws-s3
</p>

**Comment by OrkoHunter** [comment]
@ayshiff Good that you noticed it! If you make it with draw.io and save as an svg, it will be easier to maintain/update! Try the [draw.io VS Code extension](https://marketplace.visualstudio.com/items?itemName=hediet.vscode-drawio) and open a file with `.drawio.svg` extension to experience the magic. ✨

**Comment by andrewthauer** [comment]
> @ayshiff Good that you noticed it! If you make it with draw.io and save as an svg, it will be easier to maintain/update! Try the [draw.io VS Code extension](https://marketplace.visualstudio.com/items?itemName=hediet.vscode-drawio) and open a file with `.drawio.svg` extension to experience the magic. ✨

Or `plantuml` if you'd rather not thinking about layout :smile:

File: packages/techdocs-common/package.json
Comment (andrewthauer) [inline packages/techdocs-common/package.json:45]: It looks like v3 was just released. Is it worth taking a look? Seems like it has better TS support - https://aws.amazon.com/about-aws/whats-new/2020/12/aws-sdk-javascript-version-3-generally-available/.

File: plugins/techdocs/package.json
Comment (andrewthauer) [inline plugins/techdocs/package.json:137]: It might be worth moving this out into a `schema.d.ts` file and using the `files` entry instead here.

**Review by andrewthauer — COMMENTED** [review]
Question - Will publishers support using nested directories? For instance, you want to share an s3 bucket for other purposes and the documentation sites are located in say `/docs/sites` as the root.

File: plugins/techdocs/package.json
Comment (OrkoHunter) [reply plugins/techdocs/package.json:137]: Since https://github.com/backstage/backstage/pull/3799 is doing this for all techdocs packages, I think it's okay to skip it here.

**Comment by OrkoHunter** [comment]
> Will publishers support using nested directories? For instance, you want to share an s3 bucket for other purposes and the documentation sites are located in say /docs/sites as the root.

Great question @andrewthauer and big +1 for adding that support. I'll create an issue for this. I imagine it can be defined as a `rootPath` inside `techdocs.publisher.awsS3`. Will do for both GCS and S3.

File: packages/techdocs-common/package.json
Comment (OrkoHunter) [reply packages/techdocs-common/package.json:45]: Nice! @ayshiff Would you this this a try? You can create an issue for the upgrade if you want to do it later or want someone else to do it - and that is totally fine.

Comment (ayshiff) [reply packages/techdocs-common/package.json:45]: Yes of course ! I'll look into it 👍

File: docs/features/techdocs/configuration.md
Comment (mfrinnstrom) [reply docs/features/techdocs/configuration.md]: If I run this as a container in ECS will I be able to skip all of this config and use the role that is assigned to the task?

Comment (OrkoHunter) [reply docs/features/techdocs/configuration.md]: @mfrinnstrom Role and Region are not used as of this moment. Only bucket name and credentials are needed.

Comment (mfrinnstrom) [reply docs/features/techdocs/configuration.md]: @OrkoHunter that was my concern. If I add credentials to my task in AWS ECS that would be picked up automatically by the AWS SDK. It seems like we bypass this here and require credentials to be provided from the app.

Comment (OrkoHunter) [reply docs/features/techdocs/configuration.md]: > If I add credentials to my task in AWS ECS

Is this same as setting environment variables? We are open for suggestions. When the AWS sdk is initialized here, should it check some existing environment variables?
May I also ask what are the downsides of explicitly setting those environment variables in `app-config.yaml`?

File: packages/techdocs-common/src/stages/publish/awsS3.ts
Comment (ayshiff) [inline packages/techdocs-common/src/stages/publish/awsS3.ts:155]: Do we have another way to get the `Readable | ReadableStream | Blob` in a better way ?

File: docs/features/techdocs/configuration.md
Comment (mfrinnstrom) [reply docs/features/techdocs/configuration.md]: @OrkoHunter not really, although the AWS SDK has some standard environment variables that are picked up automatically if available. Some more information can be found [here](https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/setting-credentials-node.html).

We usually create roles and assign them to our different workloads, in this case an ECS task. This will then be picked up by the AWS SDK in the ECS task without having to create a user and generate credentials for that, and then having to securely manage and supply those credentials to the workload. 

This works great when running in AWS but when running outside of AWS you will need something like what is implemented right now with the access and secret keys as environment variables. It should probably be possible to use the default environment variables (`AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`) and have the SDK pick that up automatically, but that becomes a bit hidden and maybe not to easy to follow in the config.

**Comment by ayshiff** [comment]
@OrkoHunter Do you know why the `lint` and `tsc` commands fail on the CI on a line that does not seem to exist ?

**Comment by OrkoHunter** [comment]
Hey @ayshiff! Could you try pulling the latest remote master branch into your feature branch? (Or do git rebase) That will cause some conflicts, maybe.

This https://github.com/backstage/backstage/pull/3764 was merged recently.

**Comment by OrkoHunter** [comment]
@ayshiff The failing test is gone (we had to rebase and fix merge conflict, that's all). Now there is a new `tsc` issue with the `aws-sdk` package it seems.

I also had to run `yarn` again after copying over `yarn.lock` from the master branch to sync.

**Comment by ayshiff** [comment]
Thanks @OrkoHunter 👍 I'll look into it.

**Comment by ayshiff** [comment]
I wonder why I didn't have conflicts while doing my rebase 🤔 
I was probably not up to date with the remote. My bad.

**Comment by ayshiff** [comment]
It looks like an issue the `aws-sdk-js-v3` has: [Related issue](https://github.com/aws/aws-sdk-js-v3/issues/1803)

**Comment by andrewthauer** [comment]
> It looks like an issue `aws-sdk-js-v3` has: [Related issue](https://github.com/aws/aws-sdk-js-v3/issues/1803)

That's unfortunate. Maybe not worth the version bump until that is sorted out.

**Comment by ayshiff** [comment]
I'm going to move the v3 bump commits to another branch and I'm going to create a Draft PR while waiting for the issue to be fixed in [#1812](https://github.com/aws/aws-sdk-js-v3/pull/1812).

Does it sound good to you ?

**Comment by OrkoHunter** [comment]
Hey @ayshiff, the failing CI is unrelated to this PR. Should be fixed when https://github.com/backstage/backstage/pull/3825 is merged.

**Comment by freben** [comment]
And now it is! A rebase should be sufficient.

**Comment by ayshiff** [comment]
I redid a test with version **3.1.0** of `aws-sdk-js-v3` which fixes the compiling error we had with the waiter.
However a new compilation error has appeared (https://github.com/aws/aws-sdk-js-v3/issues/1842) which should be fixed in a future PR.

I will do some testing again soon before opening a new PR to bump the sdk version and to be sure it doesn't break anything.
I just wanted to keep you informed on the status of the compiling issue resolution.

**Comment by OrkoHunter** [comment]
@ayshiff Heads up, please update PR to use the `config.d.ts` files for the newly introduced AWS configs. :) Ref: https://github.com/backstage/backstage/pull/3799

File: packages/techdocs-common/src/stages/publish/awsS3.ts
Comment (lowjoel) [inline packages/techdocs-common/src/stages/publish/awsS3.ts:34]: Possible improvement for a future iteration (happy to contribute if we reach there first): this should use either EC2 instance profiles/Kubernetes IRSA to remove the need to provide an access key. These credentials are automatically rotated and built into the AWS SDK.

Comment (mfrinnstrom) [reply packages/techdocs-common/src/stages/publish/awsS3.ts:34]: Maybe have the `credentials` as optional and if not present use the default credentials chain that @lowjoel mentions? Just `new aws.S3()` basically. 

We would also be happy to contribute this since we have a policy to never create users and use access keys. I can understand that this might be necessary if you are not running in AWS but still want to use S3 though.

Comment (ayshiff) [reply packages/techdocs-common/src/stages/publish/awsS3.ts:34]: I'll be happy to add this to the existing PR !

@OrkoHunter Do we keep this PR in current state and iterate over it once merged to add optional credentials ? Or do we add that on this branch ?

Comment (OrkoHunter) [reply packages/techdocs-common/src/stages/publish/awsS3.ts:34]: @ayshiff IMO let us iterate in the future! We shouldn't delay this PR more. Hopefully we can merge this next week. :)

**Comment by OrkoHunter** [comment]
@backstage/maintainers Please review when possible!

File: packages/techdocs-common/src/stages/publish/awsS3.ts
Comment (benjdlambert) [inline packages/techdocs-common/src/stages/publish/awsS3.ts:96]: Prefer if we could use `try` `catch` around promises  rather than encapsulating all this logic. I think it might be a little cleaner.

`fs-extra` I think will return promises for most of the operations that you require here, or you can use `fs.promises` from the standard library instead.

From a readers POV it can be super hard to read and understand without `async/await`

File: packages/techdocs-common/src/stages/publish/awsS3.ts
Comment (benjdlambert) [inline packages/techdocs-common/src/stages/publish/awsS3.ts:141]: Same here

File: packages/techdocs-common/src/stages/publish/awsS3.ts
Comment (benjdlambert) [inline packages/techdocs-common/src/stages/publish/awsS3.ts:210]: And same here

File: docs/features/techdocs/configuration.md
Comment (OrkoHunter) [reply docs/features/techdocs/configuration.md:77]: Yep! There are some additional ways to handle S3 configuration apart from creating an access user agent (discussed above in various threads). Thinking we can address it in a separate PR (to also improve the GCS auth on similar patterns).

File: app-config.yaml
Comment (OrkoHunter) [reply app-config.yaml:79]: Hmm, interesting idea of unification of configs based on provider as you write above. I think we might do this just like integrations with GitHub/GitLab/etc. ➕ 
As of now for `techdocs.publisher.type`, IMHO `googleGcs` and `awsS3` are slightly better than `gcs` and `s3`. :)

**Review by gustavohalmvia — APPROVED** [review]

**Review by benjdlambert — APPROVED** [review]

**Comment by OrkoHunter** [comment]
@ayshiff Do you think this is good to go? I'll do a final Quality Assurance on this and merge. Could you also highlight some future fixes/improvements? We can create issues for them.

**Comment by ayshiff** [comment]
I think this is good to go, I’ll do a last check in the afternoon.

Here are the future fixes/improvements that I noted in the comments:

- handle S3 configuration apart from creating an access user agent ( add the possibility to read from the instance profile or `~/.aws/credentials` ) [Link](https://github.com/backstage/backstage/pull/3794#discussion_r550371101)
- Drop provider prefixes ? (`awsS3` -> `s3` and `googleGcs` -> `gcs`) [Link](https://github.com/backstage/backstage/pull/3794#discussion_r550920073)
- Unification of configs based on provider [Link](https://github.com/backstage/backstage/pull/3794#discussion_r550920073)
- `aws-sdk`version bump (already implemented on a branch, waiting for some TypeScript issues to be fixed on `aws-sdk-js-v3`.)

**Comment by OrkoHunter** [comment]
@ayshiff Do let me know when it's ready according to you.
Thank you for summarising the improvements. I can go ahead and create separate issues for them once this is merged. But feel free and let me know if you want to create them yourself. :)

**Comment by ayshiff** [comment]
It looks good for me !
Do you want me to clean up the commit history ? (squash some commits ...)

I’ll be happy to create them. :)
For issues involving aws s3 and gcs do I create two issues or a single one that includes both ?

**Comment by OrkoHunter** [comment]
> Do you want me to clean up the commit history ? (squash some commits ...)

No, the commit history looks good to me!

> I’ll be happy to create them. :)
For issues involving aws s3 and gcs do I create two issues or a single one that includes both ?

Awesome and thank you! This is what I think -

1. (Only S3) Improved S3 credentials handling. Explicitly defining configs should not be required.
2. (Only GCS) Same ^. One should be able to do `new GCS()`, and it should not complain about missing configs. The logic should be - during initialization, if the publisher is able to read the metadata of the bucket -> Connected; unable to read metadata -> Throw error.
2. (Only S3) aws-sdk version bump (already implemented on a branch, waiting for some TypeScript issues to be fixed on aws-sdk-js-v3.)
3. (Joint) Enable publishers support using nested directories in a bucket

**Comment by OrkoHunter** [comment]
You did awesome on this PR @ayshiff! 👏 I appreciate your attention to detail with the changes you did, and the reviews you addressed (sparked many good conversations). :)

**Comment by ayshiff** [comment]
I think the problem with failed tests on master is due to bad path delimiters. 
Do we revert the PR and add that or do I create another PR that fixes the tests ?

**Comment by OrkoHunter** [comment]
@ayshiff New one would be good!

**Comment by OrkoHunter** [comment]
You could use something like this in the tests.
```
const path = os.platform() === 'win32' ? 'C:\\rootDir' : '/rootDir';
```

**Pull request merged**
